nexus/pkg/artifacts/downloader.go

```diff
@@ -0,0 +1,309 @@
+	if versionIndex == nil {
+		return fmt.Errorf("invalid artifact download, missing version index")
+	}
+	artifactDir, err := env.GetArtifactDir()
+	if err != nil {
+		return err
+	}
+	downloadRoot := filepath.Join(artifactDir, fmt.Sprintf("%s:v%d", sourceArtifactName, *versionIndex))
+
+	path := CheckExists(downloadRoot)
+	if path == nil {
+		downloadRoot = SystemPreferredPath(downloadRoot, false)
+	}
+	ad.DefaultDownloadRoot = &downloadRoot
+	return nil
+}
+
+func (ad *ArtifactDownloader) setUpstreamArtifacts(manifest Manifest) error {
+	for _, entry := range manifest.Contents {
+		referencedID, err := getReferencedID(entry.Ref)
+		if err != nil {
+			return err
+		}
+		if referencedID != nil {
+			response, err := gql.ArtifactByID(
+				ad.Ctx,
+				ad.GraphqlClient,
+				*referencedID,
+			)
+			if err != nil {
+				return err
+			} else if response == nil {
+				return fmt.Errorf("could not get artifact by id for reference %s", *entry.Ref)
+			}
+			depArtifact := response.GetArtifact()
+			if depArtifact != nil {
+				ad.UpstreamArtifacts = append(ad.UpstreamArtifacts, *depArtifact)
+			}
+		}
+	}
+	return nil
+}
+
+func (ad *ArtifactDownloader) downloadFiles(artifactID string, manifest Manifest) error {
+	var downloadRoot string
+	if ad.DownloadRoot != nil {
+		downloadRoot = *ad.DownloadRoot
+	} else if ad.DefaultDownloadRoot != nil {
+		downloadRoot = *ad.DefaultDownloadRoot
+	}
+	if downloadRoot == "" {
+		return fmt.Errorf("download root was not set")
+	}
+	// retrieve from "WANDB_ARTIFACT_FETCH_FILE_URL_BATCH_SIZE"?
+	batchSize := BATCH_SIZE
+
+	type TaskResult struct {
+		Task *filetransfer.Task
+		Name string
+	}
+
+	// Fetch URLs and download files in batches
+	manifestEntries := manifest.Contents
+	numInProgress, numDone := 0, 0
+	nameToScheduledTime := map[string]time.Time{}
+	taskResultsChan := make(chan TaskResult)
+	manifestEntriesBatch := make([]ManifestEntry, 0, batchSize)
+
+	for numDone < len(manifestEntries) {
+		cursor := ""
+		hasNextPage := true
+		for hasNextPage {
+			// Prepare a batch
+			now := time.Now()
+			manifestEntriesBatch = manifestEntriesBatch[:0]
+			response, err := gql.ArtifactFileURLs(
+				ad.Ctx,
+				ad.GraphqlClient,
+				artifactID,
+				&cursor,
+				&batchSize,
+			)
+			if err != nil {
+				return err
+			} else if response == nil {
+				return fmt.Errorf("could not fetch artifact file urls for %s", ad.QualifiedName)
+			}
+			hasNextPage = response.GetArtifact().GetFiles().PageInfo.HasNextPage
+			cursor = *response.GetArtifact().GetFiles().PageInfo.EndCursor
+			for _, edge := range response.GetArtifact().GetFiles().Edges {
+				filePath := edge.GetNode().Name
+				entry, err := manifest.GetManifestEntryFromArtifactFilePath(filePath)
+				if err != nil {
+					return err
+				}
+				if entry.Ref != nil {
+					numDone++
+					continue
+				}
+				entry.DownloadURL = &edge.GetNode().DirectUrl
+				entry.LocalPath = &filePath
+				if _, ok := nameToScheduledTime[*entry.LocalPath]; ok {
```

Comment:
Move this below line 224. Otherwise
- you do a bunch of work that might not be necessary
- you modify `entry` while it's being used by other goroutines, which is not thread-safe